api/types/types.go

```diff
@@ -524,14 +524,12 @@ type Runtime struct {
 	Path string   `json:"path"`
 	Args []string `json:"runtimeArgs,omitempty"`
 
-	// This is exposed here only for internal use
-	// It is not currently supported to specify custom shim configs
-	Shim *ShimConfig `json:"-"`
+	Shim *ShimConfig `json:"shim,omitempty"`
 }
 
 // ShimConfig is used by runtime to configure containerd shims
 type ShimConfig struct {
-	Binary string
+	Binary string      `json:"binary,omitempty`
 	Opts   interface{}
```

Comment:
Maybe a `*ptypes.Any` here.
Or perhaps define our own equivalent w/o importing proto libs which we can convert internally.

Reply:
Hello @cpuguy83 -, I'm not sure why we need proto type here? How we used it for? Is it enough to transfer "shim" info through grpc by just seting the binary type to "ptypes.Any"?

Reply:
Right, we don't want to handle the type in dockerd, especially since it's just passed off to shim anyway.

The `*Any` type is just a proper encoding of `interface{}`.

Reply:
Hello @cpuguy83 -, I twist the code again, now it works well.

Reply:
I'm not sure I understand the change you added.

What we should be able to do is:

1. Tell containerd which shim to use
    - Note that containerd only currently supports resolving a reference name like `io.containerd.runc.v2` to `containerd-shim-runc-v2`, it does not support custom paths
2. Pass along a config to the shim on create.
    - This is specific to each shim, eg `runctypes.RuncOptions`

To do 2, we either need to have explicit knowledge of every shim we ever want to support OR we should just be passing through configs.
In order to pass through those configs we need a place for users to encode them.
In this case we are looking to do that in the daemon config that gets stored in `/etc/docker/daemon.json`.
An `interface{}` does allow us to do this, but then we are passing around untyped data, and the containerd API is expecting to fill a `types.Any` (from the protobuf lib).
A `types.Any` has a type URL (i.e. full path of type... e.g. `github.com/my/shim.Options`) and some data.
https://github.com/gogo/protobuf/blob/master/protobuf/google/protobuf/any.proto

To understand how this is consumed, check out https://github.com/containerd/typeurl

These options get passed along to shims when containerd calls `Create` against the shim.

Reply:
Sorry for my poor code. I have never touched these proto things before. Thanks for your guidance, I begin to aware how to deal with these. Really appreciated @cpuguy83

Reply:
Hello @cpuguy83 -, I set shim.opts to types.Any and keep binary as string, thus let opts to convey the config. Is it right?